src/txn/txn.c

```diff
@@ -769,10 +770,10 @@ __txn_append_hs_record(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_PAGE *
     *fix_updp = upd;
 
     /*
-     * When the prepared update is getting committed, use the retrieved history store update to form
-     * a proper stop timestamp in the history store.
+     * When the prepared update is getting committed or the history store update is still on the
+     * update chain, no need to append it onto the update chain.
      */
-    if (commit)
+    if (commit || !prepare_on_disk)
         goto done;
```

Comment:
Use the `first_committed_upd_in_hs` indicates that the hs update exists in the update list instead of `prepare_on_disk`. Both of them should represent the same, but that depends on the eviction failure location.